### PR TITLE
[FLINK-4458] Replace ForkableFlinkMiniCluster by LocalFlinkMiniCluster

### PR DESCRIPTION
**Pull request by tillrohrmann**
This PR is based on #2449. Thus, only the commits 5a2c70f and e07ad44 are relevant.

The PR refactors the `LocalFlinkMiniCluster` and the `TestingCluster` so that they can replace the `ForkableFlinkMiniCluster`. The `LocalFlinkMiniCluster` now starts all distributed components under a random port. This is possible with #2449.

The difference of the `TestingCluster` compared to the `LocalFlinkMiniCluster` is that it starts the testing components (e.g. `TestingJobManager`, `TestingTaskManager`, etc.) instead of the plain ones. It also allows to start the `JobManager` in a synchronous mode. Furthermore, it offers some operations like restarting the leading `JobManager` under the same port.

### CODE REVIEW

File: flink-runtime/src/main/scala/org/apache/flink/runtime/minicluster/FlinkMiniCluster.scala [hunk 1]
```diff
@@ -69,7 +69,7 @@ abstract class FlinkMiniCluster(
     ConfigConstants.JOB_MANAGER_IPC_ADDRESS_KEY,
     InetAddress.getByName("localhost").getHostAddress())
 
-  val configuration = generateConfiguration(userConfiguration)
+  protected val _configuration = generateConfiguration(userConfiguration)
```
Comment (StephanEwen) [inline flink-runtime/src/main/scala/org/apache/flink/runtime/minicluster/FlinkMiniCluster.scala]: Why the underscore here?

**Comment by StephanEwen** [comment]
I think these changes are good.

As a very personal and biased statement: I think the code overdoes it a bit with the "every parameter on a new line" policy. We have beautiful wide screens and I get sore fingers from scrolling up and down again to figure out enough context of a code passage ;-)

Also, the style breaks the "visual" separation between different parts of statement, like "members of a return tuple" vs. "function parameters", or "clauses of an if statement" and "body of the if statement". I know the idea of that pattern was to improve code readability, but I am getting the feeling this is approaching a "verschlimmbesserung" (an improvement that makes things worse).

File: flink-runtime/src/main/scala/org/apache/flink/runtime/minicluster/FlinkMiniCluster.scala (hunk 1 above)
Comment (tillrohrmann) [inline flink-runtime/src/main/scala/org/apache/flink/runtime/minicluster/FlinkMiniCluster.scala]: It's actually the original configuration which will usually have the job manager ipc port set to 0. Thus, in order to return a configuration which points to the current leader, we have to generate a new configuration via `configuration` which inserts the selected port. I'll rename the variable to `originalConfiguration`.

**Comment by tillrohrmann** [comment]
Thanks for the review @StephanEwen.

I agree that we've different style preferences concerning method parameters. I also agree that the one parameter per line is a little bit more verbose and requires more vertical scrolling. The best approach would probably be to reduce the number of method parameters by grouping them.

**Comment by tillrohrmann** [comment]
Will rebase on the latest master and #2449 to run the travis tests again.

**Comment by tillrohrmann** [comment]
Builds are passing on Travis (own repository). Would like to merge this PR if there are no objections.

**Comment by StephanEwen** [comment]
Quick question: The `ForkableFlinkMiniCluster` was something semi-public, in the sense that it was part of the `flink-test-utils` project and used by quite a few users. So this would be a breaking change.

**Comment by tillrohrmann** [comment]
Yes that is true. However, the `flink-test-utils` have not been guaranteed to be stable if I'm not mistaken. The changes one has to apply is to replace the `ForkableFlinkMiniCluster` with the `LocalFlinkMiniCluster`.

**Comment by tillrohrmann** [comment]
I will re-introduce the `ForkableFlinkMiniCluster` which is simply extending the `LocalFlinkMiniCluster`.

**Comment by tillrohrmann** [comment]
I've re-introduced the `ForkableFlinkMiniCluster`. However, this time it only starts the normal distributed component classes (e.g. `JobManager` instead of `TestingJobManager`, etc.). This implies that the testing messages are no longer supported when using the `ForkableFlinkMiniCluster`.

I think the user should not meddle with the testing components in the first place. Therefore, it seems to be fair to offer the user a mini cluster to run his tests but not one which instantiates testing components. The recommended cluster is the `LocalFlinkMiniCluster` from now on. However, in order to not break all existing test code, we still maintain the `ForkableFlinkMiniCluster`.

Not having to start the testing classes allows to move them back into the test scope of `flink-runtime`, since they are no longer required by the `flink-test-utils` module.

**Comment by tillrohrmann** [comment]
Rebasing on the latest master and letting Travis run. If it gives green light, I will merge the PR.

**Pull request closed without merging**